Add support for position:sticky

mrobinson opened the pull request:
We add a new clip scroll tree node type which is a sticky frame. The
StickyFrame doesn't do any clipping, but will adjust the position of its
contents to satisfy a given set of sticky constraints. The constraints
consist of a margin (distance from the particular edge of the viewport)
and a max_offset, which determines how far the sticky frame can move
from its original (scrolled) position. The design of these  constraints
can be modified if they are particularly unpleasant for Gecko to
implement.

Fixes #1277.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/1573)
<!-- Reviewable:end -->

mrobinson commented:
The servo part of this patch is here: https://github.com/mrobinson/servo/tree/position-sticky

There is still one test failing in Servo, but I'm almost certain it is unrelated to the WebRender changes.

glennw commented:
Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[webrender/src/clip_scroll_node.rs, line 68 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXzCZ8wjAr6NGNH7TM:-KrXzCZ8wjAr6NGNH7TN:b-cnhv3i) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_node.rs#L68)):*
> ```Rust
>     /// Transform for this layer, relative to parent reference frame. A reference
>     /// frame establishes a new coordinate space in the tree.
>     ReferenceFrame(LayerToScrollTransform, LayerVector2D),
> ```

Can we add a comment here describing this vector, or ideally, switch this to an enum with named fields?

---

*[webrender/src/clip_scroll_node.rs, line 77 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXzQws0DqBraAKWNMr:-KrXzQwt4gqLTwNe8QAb:bez72tc) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_node.rs#L77)):*
> ```Rust
>     ScrollFrame(ScrollingState),
> 
>     /// A special kind of node that adjusts its position based on the position
> ```

Perhaps worth including a link to the specification describing how position: sticky works?

---

*[webrender/src/clip_scroll_tree.rs, line 40 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrY-p_7khDJ3pp0L6Ue:-KrY-p_8emwHCFBaM5MP:b-3zpga5) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_tree.rs#L40)):*
> ```Rust
> }
> 
> pub struct TransformUpdateState {
> ```

Nice change!

---

*[webrender/src/clip_scroll_tree.rs, line 400 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrY0-kAxQW3mp-ejkqK:-KrY0-kAxQW3mp-ejkqL:bxcipdw) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_tree.rs#L400)):*
> ```Rust
>             NodeType::StickyFrame(constrant_rect) => {
>                 pt.new_level(format!("StickyFrame"));
>                 pt.add_item(format!("constraint rect: {:?}", constrant_rect));
> ```

nit: variable spelling

---

*[webrender/src/frame.rs, line 454 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrY-TtgyanvGz6rztJq:-KrY-TtgyanvGz6rztJr:b-1lnwvn) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/frame.rs#L454)):*
> ```Rust
>         }
> 
>         //if let ScrollPolicy::Sticky(..) = stacking_context.scroll_policy {
> ```

Let's remove this if not needed.

---

*[webrender_api/src/display_item.rs, line 90 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXyl7fhyzWW75eatpf:-KrXyl7g8Ml_WYWcNeBh:b1c2y87) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender_api/src/display_item.rs#L90)):*
> ```Rust
> pub struct StickyFrameInfo {
>     pub left: Option<StickyInfo>,
>     pub right:Option< StickyInfo>,
> ```

nit: spaces

---

*[wrench/simple.yaml, line 1 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXyZ5cQmOYjuAi8-Gy:-KrXyZ5dyQoHNP32wq_q:bffram1) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/wrench/simple.yaml#L1)):*
> ```YAML
> root:
> ```

Is this file meant to be committed? It seems like the wrong location, at least.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1573)*
<!-- Sent from Reviewable.io -->

glennw commented:
Nice! Mostly just nits to fix up, and working out what the failing test is. Does this result in some new passes in WPT/CSS tests?

mrobinson commented:
Review status: 6 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[webrender/src/clip_scroll_node.rs, line 68 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXzCZ8wjAr6NGNH7TM:-KrZiMpRvOWKJ2bJ_U7e:b-896fix) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_node.rs#L68)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Can we add a comment here describing this vector, or ideally, switch this to an enum with named fields?
</blockquote></details>

Done.

---

*[webrender/src/clip_scroll_node.rs, line 77 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXzQws0DqBraAKWNMr:-KrZiQQJe_9sSI6WDQnX:b-896fix) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_node.rs#L77)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Perhaps worth including a link to the specification describing how position: sticky works?
</blockquote></details>

Done.

---

*[webrender/src/clip_scroll_tree.rs, line 400 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrY0-kAxQW3mp-ejkqK:-KrZiSdcNOvS9s8hxDlq:b-k0drn3) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/clip_scroll_tree.rs#L400)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: variable spelling
</blockquote></details>

Done. Nice catch! This should have been called `sticky_frame_info' anyhow. 

---

*[webrender/src/frame.rs, line 454 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrY-TtgyanvGz6rztJq:-KrZiaFF4vVhG7e0wdJv:b-896fix) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender/src/frame.rs#L454)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Let's remove this if not needed.
</blockquote></details>

Done.

---

*[webrender_api/src/display_item.rs, line 90 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXyl7fhyzWW75eatpf:-KrZiauOkktqwKEeF0_2:b-896fix) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/webrender_api/src/display_item.rs#L90)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

nit: spaces
</blockquote></details>

Done.

---

*[wrench/simple.yaml, line 1 at r1](https://reviewable.io:443/reviews/servo/webrender/1573#-KrXyZ5cQmOYjuAi8-Gy:-KrZiboJUB9CIrPI_Odq:b25nrah) ([raw file](https://github.com/servo/webrender/blob/6d0adebcdd47811e8852069c1ec5461681ca4f6a/wrench/simple.yaml#L1)):*
<details><summary><i>Previously, glennw (Glenn Watson) wrote…</i></summary><blockquote>

Is this file meant to be committed? It seems like the wrong location, at least.
</blockquote></details>

Done. I accidentally included this in the patch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/webrender/1573)*
<!-- Sent from Reviewable.io -->

mrobinson commented:
Thanks for the review! I think that I've addressed all your comments. I'm not sure if there are any tests for position:sticky in the CSS test suite. :(

mrobinson commented:
The latest version of the Servo patch fixes the test failure.

kvark commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -275,6 +321,44 @@ impl ClipScrollNode {
+            sticky_offset.y = sticky_offset.y.max(0.0).min(info.max_offset);
+        }
+
+        if sticky_offset.y == 0.0  {
```
couldn't this be just
```rust
sticky_offset.y = if let Some(info) = sticky_frame_info.top {
...
} else if let Some(info) = sticky_frame_info.bottom {
...
};
```

mrobinson replied:
We can't do this here because if, for some reason, a combination of constraint and scroll position sends the `top` offset to 0, we still want to apply the bottom constraint.

kvark commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -470,3 +554,15 @@ impl ScrollingState {
+    /// The original, untransformed and relative to the parent reference frame, origin of this
+    /// reference frame. This is already rolled into the `transform' property, but we also store it
+    /// here to properly transform the viewport for sticky positioning.
+    pub original_origin: LayerVector2D,
```
perhaps, `untransformed_origin`?

mrobinson replied:
Renamed this one. See above.

kvark commented on webrender/src/clip_scroll_tree.rs:
```diff
@@ -226,68 +235,65 @@ impl ClipScrollTree {
+                        state.nearest_scrolling_ancestor_viewport.translate(&info.original_origin),
+                },
+                NodeType::Clip(..) | NodeType::StickyFrame(..) => TransformUpdateState {
+                    parent_reference_frame_transform: state.parent_reference_frame_transform,
```
we could shortcut it to `state`:
```rust
TransformUpdateState {
parent_reference_frame_transform:...,
parent_scroll_offset: ...,
.. state
}
```

mrobinson replied:
I didn't know about this shorthand! I've used it everywhere here now.

kvark commented on webrender_api/src/display_item.rs:
```diff
@@ -78,6 +79,26 @@ pub struct ClipDisplayItem {
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+pub struct StickyFrameInfo {
```
Could use `euclid::TypedSideOffsets2D`

mrobinson replied:
Done.

kvark commented on webrender_api/src/display_item.rs:
```diff
@@ -78,6 +79,26 @@ pub struct ClipDisplayItem {
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
+pub struct StickyInfo {
```
perhaps, rename to `StickySide`?

mrobinson replied:
Renamed this to StickySideConstraint.

kvark commented on wrench/reftests/scrolling/reftest.list:
```diff
@@ -8,4 +8,4 @@
 == scroll-layer.yaml scroll-layer-ref.yaml
 == simple.yaml simple-ref.yaml
 == clip-and-scroll-property.yaml clip-and-scroll-property-ref.yaml
-== translate-nested.yaml translate-nested-ref.yaml
```
why removing the old yaml from here but not the file itself?

mrobinson replied:
This was a merge issue. Thanks for pointing it out. The line should be restored now.

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -159,9 +167,15 @@ impl ClipScrollNode {
                                content_size: LayerSize,
-                               local_transform: &LayerToScrollTransform,
+                               transform: &LayerToScrollTransform,
+                               original_origin: LayerVector2D,
```
I've renamed this to origin_in_parent_reference_frame. The "untransformed" description in my comment wasn't really accurate, so I've updated the comment as well.

mrobinson commented on webrender/src/clip_scroll_node.rs:
```diff
@@ -231,41 +264,54 @@ impl ClipScrollNode {
+                                                 &state.nearest_scrolling_ancestor_viewport);
+
+                let scrolled_parent_combined_clip =
+                    scrolled_parent_combined_clip.translate(&-sticky_offset);
```
Fixed.

kvark commented:
Awesome, thanks @mrobinson !
@bors-servo r+

bors-servo commented:
:pushpin: Commit 978b887 has been approved by `kvark`

<!-- @bors-servo r=kvark 978b88718a2af8ee9506dc9b3641261605c003e9 -->

bors-servo commented:
:hourglass: Testing commit 978b88718a2af8ee9506dc9b3641261605c003e9 with merge dec8aa06d10f218f6641e9fcfd4470cb306fc5e0...

bors-servo commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/servo/webrender/builds/265182677?utm_source=github_status&utm_medium=notification)

kvark commented:
Looks like an ICE when building Angle, again.
@bors-servo retry

mrobinson commented:
This seems to have failed building angle:
```
c++: internal compiler error: Killed (program cc1plus)
```
I'm going to retry it.

mrobinson commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit 978b88718a2af8ee9506dc9b3641261605c003e9 with merge 9d3a9bd486d1efd61b762ea25f9b902ee90b7d23...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/webrender/builds/265265376?utm_source=github_status&utm_medium=notification)
Approved by: kvark
Pushing 9d3a9bd486d1efd61b762ea25f9b902ee90b7d23 to master...
